googler

```diff
@@ -740,313 +1324,316 @@ class Result:
-        for r in results:
-            urlindex.update(r.print_entry())
-        if filtered:
-            printerr('** Enter "unfilter" to show similar results Google omitted.')
-    elif output_format == 1: # Json output
-        results_object = [r.json_object() for r in results]
-        print(json.dumps(results_object, indent=2, sort_keys=True, ensure_ascii=False))
-
-    return results, urlindex, filtered
-
-
-def read_next_command():
-    """Show omniprompt and read user command."""
-
-    global colorize, colors
-
-    message = 'googler (? for help)'
-    prompt = (colors.prompt + message + colors.reset + ' ') if colorize else (message + ': ')
-    try:
-        return input(prompt)
-    except EOFError:
-        quit_program(exitcode=0)
-
-
-def construct_search_url(baseurl, keywords, site=None):
-    """Construct the search URL with keywords and optionally a site.
-
-    baseurl should already contain queries such as start and num. This
-    function only adds the q query.
-
-    keywords should either be a list of keywords, or a single
-    space-delimited string of keywords.
+    Command line interpreter and executor class for googler.
+
+    Inspired by PSL cmd.Cmd.
+
+    Parameters
+    ----------
+    opts : argparse.Namespace
+        Options and/or arguments.
+
+    Attributes
+    ----------
+    options : argparse.Namespace
+        Options that are currently in effect. Read-only attribute.
+    keywords : str or list or strs
+        Current keywords. Read-only attribute
+
+    Methods
+    -------
+    fetch()
+    display_results(prelude='\n', json_output=False)
+    fetch_and_display(prelude='\n', json_output=False, interactive=True)
+    read_next_command()
+    help()
+    cmdloop()
     """
-    url = baseurl
-    if isinstance(keywords, list):
-        url += 'q=' + '+'.join([urllib.parse.quote_plus(kw) for kw in keywords])
-    else:
-        url += 'q=' + urllib.parse.quote_plus(keywords)
-    if site:
-        url += '+site:' + urllib.parse.quote_plus(site)
-    debugp('Search URL [%s : %s]' % (server, url))
-    return url
 
+    # Class variables
+    colors = None
 
-def open_url(url):
-    _stderr = os.dup(2)
-    os.close(2)
-    _stdout = os.dup(1)
-    os.close(1)
-    fd = os.open(os.devnull, os.O_RDWR)
-    os.dup2(fd, 2)
-    os.dup2(fd, 1)
-    try:
-        webbrowser.open(url)
-    finally:
-        os.close(fd)
-        os.dup2(_stderr, 2)
-        os.dup2(_stdout, 1)
-
+    def __init__(self, opts):
+        super().__init__()
 
-# Messaging wrappers
+        self._opts = opts
 
-def message(msg, level='INFO', ansi=None):
-    """Print message, led with [%(level)s] where level is a string, to
-    stderr.  If ansi is not None and the global variable colorize is True,
-    then print ansi at the beginning, and close with \x1b[0m.
-    """
+        self._google_url = GoogleUrl(opts)
+        self._conn = GoogleConnection(self._google_url.hostname)
+        atexit.register(self._conn.close)
```

Comment:
> Does this mean we are closing the connection after 10 fetches (each page)?

No. This is an at exit handler. Only called when program exits. And we only have one instance of `GoogleConnection` throughout the program, updated through `new_connection` and `renew_connection`.

> Also, I don't think importing a library (atexit) for a single function is required.

Or you need to have global variables and another function, instead of the standard `sys.exit`. Like the old `quit_program`. The more cross-referencing you have, the less readable and maintainable the program becomes. Therefore, registering the at exit handler at the place where it is defined is really good.

Also, most PSL module imports are extremely fast. The time to import one module is usually negligible compared to the startup time of the interpreter myself. (I believe there are exceptions, e.g., `re`, which is kind of expensive.)